DB ORM fails to compile

**Describe the bug**
DB ORM fails to compile

**To Reproduce**
Steps to reproduce the behavior:
1. Use `execSqlAsync`
2. Compile

**Expected behavior**
Code to compile just fine

**Desktop (please complete the following information):**
 - OS: Debian Sid
 - Browser N/A
 - Version master (`f522d2d70efeb5e1e7e027322e4ca1379322de6b`)

**Additional context**
Compile log: https://gist.github.com/niansa/2440c4ce20c0b2ddd143f92df2c90b37

Full code: https://gist.github.com/niansa/58ea0647548f20e8377caffb8a46c253

Probably most significant message:

    /usr/local/include/drogon/orm/SqlBinder.h:339:37: error: ‘isExceptCallback’ is not a member of ‘drogon::orm::internal::FunctionTraits<main()::<lambda(const HttpRequestPtr&, std::function<void(const std::shared_ptr<drogon::HttpResponse>&)>&&, const string&)>::<lambda(const drogon::orm::Result&)>::<lambda(const drogon::orm::Result&)>::<lambda(...)> >’

--- a/orm_lib/inc/drogon/orm/SqlBinder.h
+++ b/orm_lib/inc/drogon/orm/SqlBinder.h
@@ -323,7 +323,8 @@ class DROGON_EXPORT SqlBinder : public trantor::NonCopyable
     SqlBinder &operator=(SqlBinder &&that) = delete;
     ~SqlBinder();
     template <typename CallbackType,
-              typename traits = FunctionTraits<std::decay_t<CallbackType>>>
+              typename traits =
+                  FunctionTraits<typename std::decay<CallbackType>::type>>
     typename std::enable_if<traits::isExceptCallback && traits::isPtr,
                             self>::type &
     operator>>(CallbackType &&callback)
@@ -335,7 +336,8 @@ class DROGON_EXPORT SqlBinder : public trantor::NonCopyable
     }
 
     template <typename CallbackType,
-              typename traits = FunctionTraits<std::decay_t<CallbackType>>>
+              typename traits =
+                  FunctionTraits<typename std::decay<CallbackType>::type>>
     typename std::enable_if<traits::isExceptCallback && !traits::isPtr,
                             self>::type &
     operator>>(CallbackType &&callback)
@@ -346,12 +348,13 @@ class DROGON_EXPORT SqlBinder : public trantor::NonCopyable
     }
 
     template <typename CallbackType,
-              typename traits = FunctionTraits<std::decay_t<CallbackType>>>
+              typename traits =
+                  FunctionTraits<typename std::decay<CallbackType>::type>>
     typename std::enable_if<traits::isSqlCallback, self>::type &operator>>(
         CallbackType &&callback)
     {
         callbackHolder_ = std::shared_ptr<CallbackHolderBase>(
-            new CallbackHolder<std::decay_t<CallbackType>>(
+            new CallbackHolder<typename std::decay<CallbackType>::type>(
                 std::forward<CallbackType>(callback)));
         return *this;
     }

--- a/orm_lib/inc/drogon/orm/SqlBinder.h
+++ b/orm_lib/inc/drogon/orm/SqlBinder.h
@@ -323,7 +323,7 @@ class DROGON_EXPORT SqlBinder : public trantor::NonCopyable
     SqlBinder &operator=(SqlBinder &&that) = delete;
     ~SqlBinder();
     template <typename CallbackType,
-              typename traits = FunctionTraits<CallbackType>>
+              typename traits = FunctionTraits<std::decay_t<CallbackType>>>
     typename std::enable_if<traits::isExceptCallback && traits::isPtr,
                             self>::type &
     operator>>(CallbackType &&callback)
@@ -335,7 +335,7 @@ class DROGON_EXPORT SqlBinder : public trantor::NonCopyable
     }
 
     template <typename CallbackType,
-              typename traits = FunctionTraits<CallbackType>>
+              typename traits = FunctionTraits<std::decay_t<CallbackType>>>
     typename std::enable_if<traits::isExceptCallback && !traits::isPtr,
                             self>::type &
     operator>>(CallbackType &&callback)
@@ -346,7 +346,7 @@ class DROGON_EXPORT SqlBinder : public trantor::NonCopyable
     }
 
     template <typename CallbackType,
-              typename traits = FunctionTraits<CallbackType>>
+              typename traits = FunctionTraits<std::decay_t<CallbackType>>>
     typename std::enable_if<traits::isSqlCallback, self>::type &operator>>(
         CallbackType &&callback)
     {

--- a/orm_lib/inc/drogon/orm/SqlBinder.h
+++ b/orm_lib/inc/drogon/orm/SqlBinder.h
@@ -176,8 +176,8 @@ class CallbackHolder : public CallbackHolderBase
         run(result);
     }
 
-    CallbackHolder(Function &&function)
-        : function_(std::forward<Function>(function))
+    template <typename T>
+    CallbackHolder(T &&function) : function_(std::forward<T>(function))
     {
         static_assert(traits::isSqlCallback,
                       "Your sql callback function type is wrong!");
@@ -351,7 +351,7 @@ class DROGON_EXPORT SqlBinder : public trantor::NonCopyable
         CallbackType &&callback)
     {
         callbackHolder_ = std::shared_ptr<CallbackHolderBase>(
-            new CallbackHolder<CallbackType>(
+            new CallbackHolder<std::decay_t<CallbackType>>(
                 std::forward<CallbackType>(callback)));
         return *this;
     }

--- a/orm_lib/tests/db_test.cc
+++ b/orm_lib/tests/db_test.cc
@@ -32,6 +32,9 @@
 using namespace std::chrono_literals;
 using namespace drogon::orm;
 
+void expFunction(const DrogonDbException &e)
+{
+}
 #if USE_POSTGRESQL
 DbClientPtr postgreClient;
 DROGON_TEST(PostgreTest)
@@ -40,6 +43,12 @@ DROGON_TEST(PostgreTest)
     // Prepare the test environment
     *clientPtr << "DROP TABLE IF EXISTS USERS" >> [TEST_CTX](const Result &r) {
         SUCCESS();
+        clientPtr->execSqlAsync(
+            "select 1 as result",
+            [TEST_CTX](const drogon::orm::Result &r) {
+                MANDATE(r.size() == 1);
+            },
+            expFunction);
     } >> [TEST_CTX](const DrogonDbException &e) {
         FAULT("postgresql - Prepare the test environment(0) what():" +
               std::string(e.base().what()));
@@ -674,6 +683,7 @@ DROGON_TEST(PostgreTest)
         }
     };
     drogon::sync_wait(coro_test());
+
 #endif
 }
 #endif
@@ -686,7 +696,15 @@ DROGON_TEST(MySQLTest)
     REQUIRE(clientPtr != nullptr);
     // Prepare the test environment
     *clientPtr << "CREATE DATABASE IF NOT EXISTS drogonTestMysql" >>
-        [TEST_CTX](const Result &r) { SUCCESS(); } >>
+        [TEST_CTX](const Result &r) {
+            SUCCESS();
+            clientPtr->execSqlAsync(
+                "select 1 as result",
+                [TEST_CTX](const drogon::orm::Result &r) {
+                    MANDATE(r.size() == 1);
+                },
+                expFunction);
+        } >>
         [TEST_CTX](const DrogonDbException &e) {
             FAULT("mysql - Prepare the test environment(0) what():" +
                   std::string(e.base().what()));
@@ -1260,6 +1278,7 @@ DROGON_TEST(MySQLTest)
         }
     };
     drogon::sync_wait(coro_test());
+
 #endif
 }
 #endif
@@ -1274,6 +1293,12 @@ DROGON_TEST(SQLite3Test)
     // Prepare the test environment
     *clientPtr << "DROP TABLE IF EXISTS users" >> [TEST_CTX](const Result &r) {
         SUCCESS();
+        clientPtr->execSqlAsync(
+            "select 1 as result",
+            [TEST_CTX](const drogon::orm::Result &r) {
+                MANDATE(r.size() == 1);
+            },
+            expFunction);
     } >> [TEST_CTX](const DrogonDbException &e) {
         FAULT("sqlite3 - Prepare the test environment(0):  what():" +
               std::string(e.base().what()));
@@ -1906,6 +1931,7 @@ DROGON_TEST(SQLite3Test)
             trantor::EventLoop::getEventLoopOfCurrentThread(), 1.0s);
     };
     drogon::sync_wait(coro_test());
+
 #endif
 }
 #endif

